src/telemetry.ts

```diff
@@ -78,60 +55,146 @@ export class TelemetryWorker {
     this._reporter = new TelemetryReporter(
         packageInfo.name, packageInfo.version, packageInfo.aiKey);
   }
-}
 
-// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License.
+  static getInstance(context: vscode.ExtensionContext): TelemetryWorker {
+    if (!TelemetryWorker._instance) {
+      TelemetryWorker._instance = new TelemetryWorker(context);
+    }
+    return TelemetryWorker._instance;
+  }
 
-export async function callWithTelemetry(
-    eventName: string, outputChannel: vscode.OutputChannel,
-    enableSurvey: boolean, context: vscode.ExtensionContext,
-    callback: (
-        context: vscode.ExtensionContext, outputChannel: vscode.OutputChannel,
-        // tslint:disable-next-line:no-any
-        telemetrycontext: TelemetryContext, ...args: any[]) => any,
-    // tslint:disable-next-line:no-any
-    additionalProperties?: {[key: string]: string},
-    // tslint:disable-next-line:no-any
-    ...args: any[]): Promise<any> {
-  const start: number = Date.now();
-  const properties:
-      TelemetryProperties = {result: 'Succeeded', error: '', errorMessage: ''};
-
-  if (additionalProperties) {
-    for (const key of Object.keys(additionalProperties)) {
-      if (!properties.hasOwnProperty(key)) {
-        properties[key] = additionalProperties[key];
-      }
+  /**
+   * check if it is Microsoft internal user
+   */
+  private static isInternalUser(): boolean {
+    const userDomain: string = process.env.USERDNSDOMAIN ?
+        process.env.USERDNSDOMAIN.toLowerCase() :
+        '';
+    return userDomain.endsWith('microsoft.com');
+  }
+
+  /**
+   * Create telemetry context
+   */
+  createContext(): TelemetryContext {
+    const context: TelemetryContext = {properties: {}, measurements: {}};
+    context.properties.result = TelemetryResult.Succeeded;
+    context.properties.isInternal = TelemetryWorker.isInternalUser.toString();
+    if (this._extensionContext) {
+      context.properties.developEnvironment =
+          RemoteExtension.isRemote(this._extensionContext) ?
+          DevelopEnvironment.RemoteEnv :
+          DevelopEnvironment.LocalEnv;
+    }
+    return context;
+  }
+
+  /**
+   * Send event telemetry
+   * @param eventName event name
+   * @param telemetryContext telemetry context
+   */
+  sendEvent(eventName: string, telemetryContext: TelemetryContext): void {
+    console.log(`[debug] Sending Telemetry. Event name： ${

```

Comment:
>[debug] Sending Telemetry. [](start = 17, length = 26)

Is that good to output debug info for prod release?

Reply:
Will remove it.